File: artifacts/tulipindicators/decompose/src/nvi.c
int ti_nvi_start(const double *options)
{
  (void) options;
  return 0;
}


/*** DEPENDENCIES:
***/


int ti_nvi(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *close = inputs[0];
  const double *volume = inputs[1];
  (void) options;
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (size <= ti_nvi_start(options))
  {
    return 0;
  }
  double nvi = 1000;
  output[output_idx] = nvi;
  output_idx += 1;
  int i;
  for (i = 1; i < size; i += 1)
  {
    if (volume[i] < volume[i - 1])
    {
      nvi += ((close[i] - close[i - 1]) / close[i - 1]) * nvi;
    }
    output[output_idx] = nvi;
    output_idx += 1;
  }

  assert(((&output[output_idx]) - outputs[0]) == (size - ti_nvi_start(options)));
  return 0;
}


/*** DEPENDENCIES:
int ti_nvi_start(const double *options)
{
  (void) options;
  return 0;
}


----------------------------
***/

File: artifacts/tulipindicators/translate_ablation/src/nvi.rs
use crate::*;
use rand::Rng;

pub fn ti_nvi_start(options: Option<&[f64]>) -> i32 {
    // The C function ignores the options parameter, so we do the same
    let _ = options;
    0
}
pub fn ti_nvi(size: i32, inputs: &[&[f64]], options: Option<&[f64]>, outputs: &mut [&mut [f64]]) -> i32 {
    let close = inputs[0];
    let volume = inputs[1];
    let output = &mut outputs[0];
    
    let start = ti_nvi_start(options);
    if size <= start {
        return 0;
    }
    
    let mut nvi = 1000.0;
    output[0] = nvi;
    
    for i in 1..size as usize {
        if volume[i] < volume[i - 1] {
            nvi += ((close[i] - close[i - 1]) / close[i - 1]) * nvi;
        }
        output[i] = nvi;
    }
    
    assert!((output.len() as i32) == (size - start));
    0
}
